partiql-lang/src/main/kotlin/org/partiql/lang/eval/PartiqlAstExtensions.kt

```diff
@@ -65,3 +66,137 @@ internal fun PartiqlAst.Expr.getStartingSourceLocationMeta(): SourceLocationMeta
+internal fun PartiqlAst.Expr.Id.toBindingName(): BindingName =
+    BindingName(this.name.text, this.case.toBindingCase())
+
+/** Given a set of variable references (as [BingingName]s),
+ *  retain only those that are not captured by the given [binder],
+ *  while taking case sensitivity into account. */
+internal operator fun Set<BindingName>.minus(binder: String): Set<BindingName> {
+    return this.filter { bn -> !bn.isEquivalentTo(binder) }.toSet()
+}
+
+/** Given a set of variable references (as [BingingName]s),
+ *  retain only those that are not captured by given [binders],
+ *  while taking case sensitivity into account. */
+internal operator fun Set<BindingName>.minus(binders: Set<String>): Set<BindingName> {
+    return this.filter { bn -> !binders.any { bn.isEquivalentTo(it) } }.toSet()
+}
+
+/** Free variables in an expression.
+ *  A variable (a PartiqlAst.Expr.Id) is free, unless it is in the scope of a same-named alias introduced
+ *  in FROM, LET, or GROUP BY clauses.
+ *  Assumes that the expression has been normalized/desugared with [VisitorTransforms.basicVisitorTransforms].
+ */
+internal fun PartiqlAst.Expr.freeVariables(): Set<BindingName> {
+    val visitorFold = object : PartiqlAst.VisitorFold<Set<BindingName>>() {
+        // The normal mode of operation is walk through the most AST nodes just carrying the accumulator set along.
+
+        // When a variable reference (a PartiqlAst.Expr.Id) is encountered, add it to the accumulator.
+        override fun walkExprId(node: PartiqlAst.Expr.Id, accumulator: Set<BindingName>): Set<BindingName> {
+            return accumulator + node.toBindingName()
+        }
+
+        // The invariant of each walkExprXxx(node: Xxx, accumulator) call
+        // is to add all *free* variables of `node` expression to `accumulator`.
+        // The default implementations work for all expressions except SELECT.
+
+        // Processing the SELECT involves taking care that variable occurrences bound by FROM, LET, and GROUP BY clauses
+        // do not end up in the result, if they are in the scope of those bindings.
+        override fun walkExprSelect(node: PartiqlAst.Expr.Select, accumulator: Set<BindingName>): Set<BindingName> {
+            var current = emptySet<BindingName>()
+
+            // SELECT clauses are processed in the reverse of evaluation order, since that's how variable scoping extends.
+
+            // Note: in `SELECT e AS x ...`, free variables of `e` get into the result,
+            // but `x` *does not*, due to `x` being a 'symbol', not an 'id' in the AST.
+            // (For reference, 'x' *should not* count as either free or bound because it is not a variable,
+            // it is a spec to create an attribute in a struct.)
+            current += walkProjection(node.project, emptySet())
+
+            node.limit?.let { current += walkExpr(it, emptySet()) }
+            node.offset?.let { current += walkExpr(it, emptySet()) }
```

Comment:
TODO: `LIMIT` and `OFFSET` may need special treatment. 
Say, in a `LIMIT x+5`, the variable `x` makes sense only if it is bound outside the current SELECT query (i.e., `x` is bound globally or in an outer, containing query). 
So, this occurrence of `x` should be considered free for this SELECT, and therefore be counted in its feeeVariables() set. 

Of course, this is confusing if the id `x` is also bound in FROM.  So may want to introduce a warning about that?